flask/wrappers.py

```diff
@@ -13,19 +13,110 @@
 from werkzeug.exceptions import BadRequest
 
 from . import json
-from .globals import _request_ctx_stack
+from .globals import _app_ctx_stack
 
-_missing = object()
 
+class JSONMixin(object):
+    """Common mixin for both request and response objects to provide JSON
+    parsing capabilities.
 
-def _get_data(req, cache):
-    getter = getattr(req, 'get_data', None)
-    if getter is not None:
-        return getter(cache=cache)
-    return req.data
+    .. versionadded:: 0.12
+    """
+
+    @property
+    def is_json(self):
+        """Indicates if this request/response is in JSON format or not.  By
+        default it is considered to include JSON data if the mimetype is
+        :mimetype:`application/json` or :mimetype:`application/*+json`.
 
+        .. versionadded:: 1.0
+        """
+        mt = self.mimetype
+        if mt == 'application/json':
+            return True
+        if mt.startswith('application/') and mt.endswith('+json'):
+            return True
+        return False
+
+    @property
+    def json(self):
+        """If this request/response is in JSON format then this property will
+        contain the parsed JSON data.  Otherwise it will be ``None``.
+
+        The :meth:`get_json` method should be used instead.
+        """
+        from warnings import warn
+        warn(DeprecationWarning(
```

Comment:
`is_json()` shouldn't be neccessary, as `.json` should return `None` if there is no JSON, correct?